src/app/reporting/ReportingEngine.h

```diff
@@ -0,0 +1,138 @@
+ */
+
+/**
+ *    @file
+ *      This file defines Reporting Engine for a CHIP Interaction Model
+ *
+ */
+
+#pragma once
+
+#include <app/MessageDef/ReportData.h>
+#include <app/ReadHandler.h>
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+#include <system/TLVPacketBufferBackingStore.h>
+#include <util/basic-types.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+/*
+ *  @class ReportingEngine
+ *
+ *  @brief The reporting engine is responsible for generating reports to subscriber and reader. It is able to find the intersection
+ * between the path interest set of each subscriber/reader with what has changed in the publisher data store and generate tailored
+ * reports for each subscriber/reader.
+ *
+ *         At its core, it  tries to gather and pack as much relevant attributes changes and/or events as possible into a report
+ * message before sending that to the subscriber/reader. It continues to do so until it has no more work to do. This could be due to
+ * a couple of reasons:
+ *
+ *         - Reports are in flight to the subscribers/readers
+ *         - We have exceeded the maximum number of reports that can be flight across all subscribers/readers.
+ *         - We have no more space in the packet to stuff in more data.
+ *         - We have no more dirty data to process.
+ *
+ *         Once it surmises there is no more work to be done, it returns. If all work for a read has been completed, it will
+ *         invoke a method in the ReadHandler to finish processing.
+ *
+ */
+class ReportingEngine
+{
+public:
+    /**
+     * Initializes the reporting engine. Should only be called once.
+     *
+     * @retval #CHIP_NO_ERROR On success.
+     * @retval other           Was unable to retrieve data and write it into the writer.
+     */
+    CHIP_ERROR Init();
+
+    /**
+     * Main work-horse function that executes the run-loop.
+     */
+    void Run();
+
+    /**
+     * Main work-horse function that executes the run-loop asynchronously on the CHIP thread
+     */
+    void ScheduleRun();
+
+private:
+    friend class TestReportingEngine;
+    /**
+     * Build Single Report Data including attribute changes and event data stream, and send out
+     *
+     * @param[in]    apReadHandler    A pointer to the RenaderHandler object
+     * @retval #CHIP_NO_ERROR On success.
+     * @retval other           Was unable to retrieve data and write it into the writer.
+     */
+    CHIP_ERROR BuildAndSendSingleReportData(ReadHandler * apReadHandler);
+
+    /**
+     * Send Report via ReadHandler
+     *
+     * @param[in]    apReadHandler    A pointer to the RenaderHandler object
+     * @param[in]    aPayload        A payload that has report data
+     * @retval #CHIP_NO_ERROR On success.
```

Comment:
`@retval` does not seem to tell me much more than the code. consider removing.

Reply:
done, thanks